File: get_recent_trips.py
import boto3
import json
import argparse
import time
import sys
import requests


def sortFunc(e):
	return e['timestamp']['S']


# // sortFunc()

def getLocationInfo(strProx, strApp_id, strApp_code):
	# this method takes the following parameters:
	# strProx = '41.8842,-87.6388,250'
	# strApp_id = 'cF1SE2QqJkuUgBEQHEma'
	# strApp_code = 'rKLZInnnnJDvNE5ioQIMEg'

	strUrl = 'https://reverse.geocoder.api.here.com/6.2/reversegeocode.json'
	dictHeaders = {'Content-Type': '*'}
	dictPayload = {
		'prox': strProx,
		'mode': 'retrieveAddresses',
		'maxresults': '1',
		'gen': '9',
		'app_id': strApp_id,
		'app_code': strApp_code
	}

	response = requests.get(strUrl, headers=dictHeaders, params=dictPayload)

	jsonResponse = response.json()

	return (jsonResponse['Response']['View'][0]['Result'][0]['Location']['Address'])


# // getLocationInfo()

def main():
	parser = argparse.ArgumentParser(
		description='Example: python3 get_recent_trips.py -t "cvra-demo-VehicleTripTable-abcdef" -i "cF1SE2QqJkxyzabc" -c "xyzabcJDvNE5ioQIMEg"')

	requiredNamed = parser.add_argument_group('required named arguments')
	requiredNamed.add_argument('-t', '--VehicleTripTable',
							   help='VehicleTripTable, the DynamoDB table provisioned by the Connected Vehicle Reference Archietcture',
							   required=True)
	requiredNamed.add_argument('-i', '--HereAppId',
							   help='HERE Maps app_id for determining location (register at developer.here.com)',
							   required=True)
	requiredNamed.add_argument('-c', '--HereAppCode',
							   help='HERE Maps app_code for determining location (register at developer.here.com)',
							   required=True)

	args = parser.parse_args()

	print("Using args: " + str(args))

	strVehicleTripTable = args.VehicleTripTable

	# reverse geocoding with HERE maps
	# https://developer.here.com/documentation/geocoder/topics/example-reverse-geocoding.html

	strApp_code = args.HereAppCode
	strApp_id = args.HereAppId

	strRegion = 'us-east-1'

	dynamoDbClient = boto3.client('dynamodb', region_name=strRegion)

	sys.stdout.write('Scanning trip table ' + strVehicleTripTable + '... ')

	intStartTime = int(time.time())
	response = dynamoDbClient.scan(
		TableName=strVehicleTripTable,
		Select='ALL_ATTRIBUTES'
	)

	intEndTime = int(time.time())
	intTotalTime = intEndTime - intStartTime

	print('done (' + str(intTotalTime) + 'ms).')

	listTrips = response['Items']

	# if you want to convert the response to JSON, just  use json.dumps(response) as in the following
	# strJsonResponse=json.dumps(response)
	# print(strJsonResponse)

	# sort the list by timestamp
	listTrips.sort(key=sortFunc)

	intRecordCount = json.dumps(response['Count'])
	print("Found " + str(intRecordCount) + " items in the trip table.")
	print("listItems is a " + str(type(listTrips)))

	intTripNumber = 1
	for trip in listTrips:
		strVin = str(trip['vin']['S'])
		strTripId = str(trip['trip_id']['S'])
		strTimestamp = str(trip['timestamp']['S'])
		strLongitude = str(trip['longitude']['N'])
		strLatitude = str(trip['latitude']['N'])
		strProx = strLatitude + "," + strLongitude
		floatDistance = trip['odometer']['N']
		floatFuelConsumed = trip['fuel_consumed_since_restart']['N']

		# call a method to do reverse geocoding on given lat/long
		jsonLocationInfo = getLocationInfo(strProx, strApp_id, strApp_code)

		strAddressLabel = ""
		strCity = ""
		strState = ""
		strDistrict = ""

		try:
			strAddressLabel = jsonLocationInfo['Label']
			strCity = jsonLocationInfo['City']
			strState = jsonLocationInfo['State']
			strDistrict = jsonLocationInfo['District']
		except KeyError:
			pass

		print("**** Trip " + str(intTripNumber) + " (trip_id: " + strTripId + ", VIN: " + strVin + ")")
		print("Time: " + strTimestamp + ")")
		print("Location: near " + strAddressLabel)
		print("Neighborhood: " + strDistrict)
		print("Distance: " + str(floatDistance) + " miles")
		print("Fuel consumed: " + str(floatFuelConsumed) + " gallons")
		print("All trip data: " + str(trip))
		print()
		intTripNumber = intTripNumber + 1


main()

File: AlexaSkill/GetRecentTrips/GetRecentTrips.js
'use strict';

const AWS = require('aws-sdk');
const fetch = require('node-fetch');
const args = require('yargs').argv;

function GetTripDetails(appId, appCode, tableName) {
    AWS.config.update({
        region: "us-east-1"
    });

    var docClient = new AWS.DynamoDB.DocumentClient();

    console.log("Scanning DynamoDB VehicleTripTable for trips...");

    var params = {
        TableName: tableName
    };

    docClient.scan(params, onScan);

    async function onScan(err, data) {
        if (err) {
            console.error("Unable to scan the table. Error:", JSON.stringify(err, null, 2));
            return;
        }

        // print trip info
        console.log("Scan succeeded, " + data.Count + " trips found.");

        // sort the trip by start_time, most recent first
        var trips = data.Items;
        trips.sort(function(a, b) {
            return (b.start_time > a.start_time);
        });

        // print the 3 most recent trips
        console.log("Your 3 most recent trips were: ");
        var url; // url to call HERE Maps reverse geocoder API
        var i = 0; // index of the trip
        var response; // response from HERE Maps API
        var data; // encoded response
        var locationData; // JSON response
        var latitude;
        var longitude;
        var neighborhood; // neighborhood string extracted from response

        for (i = 0; i < 3; i++) {
            latitude = trips[i].latitude;
            longitude = trips[i].longitude;

            url = "https://reverse.geocoder.api.here.com/6.2/reversegeocode.json?app_id=" + appId + "&app_code=" + appCode + "&mode=retrieveAreas&prox=" + latitude + "," + longitude;
            response = await fetch(url);
            locationData = await (response.json());
            neighborhood = locationData.Response.View[0].Result[0].Location.Address.Label;

            console.log("vin: " + trips[i].vin + ", start time: " + trips[i].start_time + ", distance: " + (trips[i].odometer).toFixed(1) + ", neighborhood: " + neighborhood);

        } // for(i)
    } // async onScan()
} // GetTripDetails()

function main() {
    console.log('GetRecentTrips v1.0');
    console.log("-----------------------------------------");

    const appId = args.appId;
    const appCode = args.appCode;
    const tableName = args.tableName;

    console.log('HERE app_id: ' + appId);
    console.log('HERE app_code: ' + appCode);
    console.log('VehicleTripTable: ' + tableName);

    GetTripDetails(appId, appCode, tableName);

} // main

main();
